10-deploying-an-application/README.md

```diff
@@ -0,0 +1,465 @@
+  $old_versions = run_task('my_app::install', [$app_servers, $db_server],
+    "Install ${version} of the application",
+    version => $version
+  ).map |$r| { $r['previous_version'] }
+
+  run_task('my_app::migrate', $db_server)
+
+  # Don't log every action on each node only important messages
+  without_default_logging() || {
+    # Expand group references or globs before iterating
+    get_targets($app_servers).each |$server| {
+
+      # Check stats and print a message to the user
+      $stats = run_task('my_app::lb', $lb_server,
+        action => 'stats',
+        backend => $instance,
+        server => $server.name,
+        _catch_errors => $force
+      ).first
+      notice("Deploying to ${server.name}, currently ${stats["status"]} with ${stats["connections"]} open connections.")
+
+      run_task('my_app::lb', $lb_server,
+        "Drain connections from ${server.name}",
+        action => 'drain',
+        backend => $instance,
+        server => $server.name,
+        _catch_errors => $force
+      )
+
+      run_task('my_app::deploy', [$server],
+        "Update application for new version",
+      )
+
+      # Verify the app server is healthy before returning it to the the load
+      # balancer.
+      $health = run_task('my_app::health_check', $lb_server,
+        "Run Healthcheck for ${server.name}",
+        target => "http://${server.name}:5000/",
+        '_catch_errors' => true).first
+
+      if $health['status'] == 'success' {
+        info("Upgrade Healthy, Returning ${server.name} to load balancer")
+      } else {
+        # Fail the plan unless the app server is healthy or this is a forced deploy
+        unless $force {
+          fail_plan("Deploy failed on app server ${server.name}: ${health.result}")
+        }
+      }
+
+      run_task('my_app::lb', $lb_server,
+        action => 'add',
+        backend => $instance,
+        server => $server.name,
+        _catch_errors => $force
+      )
+      notice("Deploy complete on ${server}.")
+    }
+  }
+
+  run_task('my_app::uninstall', [$db_server, $app_servers],
+    "Clean up old versions",
+    live_versions => $old_versions + $version,
+  )
+}
+```
+
+
+This plan can be run with the follow command. It will randomly fail 10% of the
+time when the simulated load is high.
+
+```bash
+bolt plan run my_app::deploy version=1.0.2 app_servers=app db_server=db lb_server=lb --inventoryfile ./inventory.yaml --modulepath=./modules
+```
+
+### Parameters
+
+```puppet
+plan my_app::deploy(
+  Pattern[/\d+\.\d+\.\d+/] $version,
+  TargetSpec $app_servers,
+  TargetSpec $db_server,
+  TargetSpec $lb_server,
+  String[1] $instance = 'my_app',
+  Boolean $force = false
+)
+```
+
+- `$version` is restricted to a string matching and `x.y.z` version format
+  with the Pattern Type.
+- `$app_servers`, `$db_servers`, and `$lb_server`, Three TargetSpec parameters
+   for the different tiers of the application. The TargetSpec type any String,
```

Comment:
I think this should read "The TargetSpec type can be either a String, a Target, or an Array. You can learn more about Puppet types [here](link-to-docs)".